### ISSUE TITLE
Fix setup-rust cache

### ISSUE THREAD

**Issue by Sec-ant** [biomejs/website#537]
The current cache settings of the action `setup-rust` are not working well. We've [reached the limit](https://github.com/biomejs/website/actions/caches) and the caches are not being [restored](https://github.com/biomejs/website/actions/runs/9422765503/job/25959747141#step:10:87) or [saved](https://github.com/biomejs/website/actions/runs/9422765503/job/25959747141#step:22:4) properly. This also slows down the codegen job.

I'll look into it.

**Comment by Sec-ant** [comment]
Should be fixed by #539.

**Comment by Sec-ant** [comment]
The `synchronize` workflow will update `Cargo.toml` and `Cargo.lock` when it runs the codegen job (bump the revision of biome). And the cache key is hashed against the `Cargo.lock` file. That's why every time it runs, it will save the cache with a new cache key. Thus filling up the cache storage rapidly (10 Gb limit)

**Comment by ematipico** [comment]
Do we need caching at this point?

**Comment by Sec-ant** [comment]
> Do we need caching at this point?

Yes, or rust will build all the dependencies each time we run codegen.

**Comment by Sec-ant** [comment]
After #552, we now use a single cache key `rust-cache` to restore and update the rust cache, the cache storage will no longer filled by different rust caches and we can still speed up our build time with it.

![image](https://github.com/biomejs/website/assets/10386119/7873dd1a-a29b-43ad-af2b-3972515edda3)
![image](https://github.com/biomejs/website/assets/10386119/289c7e8a-02da-4c25-8fea-758c30b6b337)

**Comment by Sec-ant** [comment]
I'm sorry for reopening this again. But the current workaround is not ideal for the following reasons:

1. The cache paths are not properly cleaned before saving, that will result in a bigger and bigger cache file. We should take the clean-up functions in [`moonrepo/setup-rust`](https://github.com/moonrepo/setup-rust/blob/dcab3dcf9f278945f3f983bace1c7c84033433a7/src/cargo.ts#L101-L153) for reference
2. The cache saving action cannot be put in a post run step, which prolongs the job duration.
3. The current workaround is not reusable.
4. We need something good enough we can use for all our repos.

To address the above issues, I plan to fork https://github.com/moonrepo/setup-rust and do some experiments.

So I reopen this issue as a reminder.

**Comment by ematipico** [comment]
CI and caching, two difficult things combined! You're brave @Sec-ant 👏

**Comment by nhedger** [comment]
@Sec-ant I'm closing this since it hasn't seen activity in a moment, and the current setup seems to work fine. Feel free to reopen if you intend to work on it.
